ffi/arcreader.lua

```diff
@@ -0,0 +1,178 @@
+--[[--
+@module ffi.arcreader
+]]
+
+local ffi = require "ffi"
+local libarchive = ffi.loadlib("archive", "13")
+require "ffi/libarchive_h"
+require "ffi/posix_h"
+
+local ArcReader = {}
+
+local ENTRY_MODE = {
+    [libarchive.AE_IFREG]  = "file",
+    [libarchive.AE_IFLNK]  = "link",
+    [libarchive.AE_IFSOCK] = "socket",
+    [libarchive.AE_IFCHR]  = "char device",
+    [libarchive.AE_IFBLK]  = "block device",
+    [libarchive.AE_IFDIR]  = "directory",
+    [libarchive.AE_IFIFO]  = "named pipe",
+}
+
+local function archive_error_string(archive)
+    local err = libarchive.archive_error_string(archive)
+    return err ~= nil and ffi.string(err) or nil
+end
+
+function ArcReader:new()
+    local o = {}
+    setmetatable(o, self)
+    self.__index = self
+    o.archive_entry = ffi.new("struct archive_entry *[1]")
+    o.entries = {}
+    o.size = 0
+    o.err = nil
+    return o
+end
+
+function ArcReader:open(filepath)
+    self.err = nil
+    if not filepath then
+        return
+    end
+    self.archive = ffi.gc(libarchive.archive_read_new(), libarchive.archive_free)
+    libarchive.archive_read_support_format_all(self.archive)
+    libarchive.archive_read_support_filter_all(self.archive)
+    if libarchive.archive_read_open_filename(self.archive, filepath, 10240) ~= libarchive.ARCHIVE_OK then
+        self.err = archive_error_string(self.archive)
+        print("ArcReader: archive_read_open_filename failed:", self.err)
+        libarchive.archive_free(self.archive)
+        ffi.gc(self.archive, nil)
+        self.archive = nil
+        return
+    end
+    self.filepath = filepath
+    self.index = 0
+    return true
+end
+
+function ArcReader:next()
+    self.err = nil
+    if self.archive == nil then
+        return
+    end
+    -- We can't consume the data twice: tweak the index
+    -- to trigger a reset in case of subsequent attempt.
+    self.index = math.floor(self.index) + 0.1
+    local err = libarchive.archive_read_next_header(self.archive, self.archive_entry)
```

Comment:
Do they have some sample C code somewhere or some such btw? Or is it entirely from scratch?

Reply:
There are some examples on the [wiki](https://github.com/libarchive/libarchive/wiki/Examples), and in some of the [manpages](https://github.com/libarchive/libarchive/wiki/ManualPages).

Reply:
Thanks.